### PR TITLE
Add Similarity query for embeddings.

### PR DESCRIPTION
**Pull request by sebpuetz**
Add a method to query for the most similar words given a query vector.

Thought, this could be nice to have since analogy queries are essentially doing the same.

### CODE REVIEW

**Comment by sebpuetz** [comment]
@danieldk 
I don't have write access to this repo, so I can't request a review.

**Comment by danieldk** [comment]
@sebpuetz I have given you owner privileges for the whole finalfusion organization.

**Comment by sebpuetz** [comment]
> 
> 
> @sebpuetz I have given you owner privileges for the whole finalfusion organization now.

Oh nice, thanks!

File: src/similarity.rs [hunk 1]
```diff
@@ -194,16 +194,36 @@ pub trait Similarity {
...
-    fn similarity(&self, word: &str, limit: usize) -> Option<Vec<WordSimilarity>>;
+    fn word_similarity(&self, word: &str, limit: usize) -> Option<Vec<WordSimilarity>>;
+
+    /// Find words that are similar to the query embedding.
+    ///
+    /// The similarity between the query embedding and other embeddings is
+    /// defined by the dot product of the embeddings. If the vectors are unit
+    /// vectors (e.g. by virtue of calling `normalize`), this is the cosine
+    /// similarity. At most, `limit` results are returned.
+    fn embedding_similarity(
```
Comment (danieldk) [inline src/similarity.rs]: I don't like overpacking traits with somewhat different functionality. Could you split this into two traits: `WordSimilarity` and `EmbeddingSimilarity`. We could consider folding `word_similarity_by` and `embedding_similarity_by` into these two respective traits.

File: src/similarity.rs [hunk 2]
```diff
@@ -445,13 +493,31 @@ mod tests {
...
+        let f = File::open("testdata/similarity.bin").unwrap();
+        let mut reader = BufReader::new(f);
+        let embeddings = Embeddings::read_word2vec_binary(&mut reader).unwrap();
+        let embedding = embeddings.embedding("Berlin").unwrap();
+        let result = embeddings.embedding_similarity(embedding.as_view(), 10);
+        assert!(result.is_some());
+        let mut result = result.unwrap().into_iter();
+        assert_eq!(10, result.len());
+        assert_eq!(result.next().unwrap().word, "Berlin");
+        println!("{:?}", result);
```
Comment (danieldk) [inline src/similarity.rs]: Stray `println!`. I know it's also in the other test, but we shouldn't introduce it in this test as well ;).

File: src/similarity.rs (hunk 1 above)
Comment (sebpuetz) [reply src/similarity.rs]: Did that, although I had to rename the `WordSimilarity` struct to `WordSimilarityResult`

**Comment by sebpuetz** [comment]
> This is a nice addition! People can probe with their custom vectors.

What do you think of adding a mask to these queries? Take a `HashSet<&str>` to allow skipping certain neighbours, rather than defaulting to an empty `HashSet`?

**Comment by sebpuetz** [comment]
Seems like my branch was coming from my fork's master branch which wasn't up to date. I'm rebasing on finalfusion/finalfusion-rust master now.

**Comment by sebpuetz** [comment]
The PR should no be mergeable without conflicts

**Comment by danieldk** [comment]
Great! Feel free to merge it if CI doesn't complain.

**Pull request merged**